### PR TITLE
Add stack to errors with the errorDetails option

### LINKED ISSUE #8425: Adding call stacks to errors
## Feature request

**What is the expected behavior?**
Adding the stack trace of mysterious errors in the `formatError` in `Stats.js`

**What is motivation or use case for adding/changing the behavior?**
I spent a whole day debugging the following error while trying to build an Angular app. 

image

Now good luck finding out what's the issue there 😄 

So I hacked around the Angular CLI and eventually reached `Stats.js` and saw that the stack property isn't getting added to the error text. 

**How should this be implemented in your opinion?**
I'll create a PR to add it

**Are you willing to work on this yourself?**
yes

### PR DESCRIPTION
**Pull request by shairez**
fixes #8425 

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
Now, whenever you'll pass `errorDetails` to `toJson()` of Stats you'll also see the error's stack trace if it exists.

### CODE REVIEW

File: test/Stats.unittest.js [hunk 1]
```diff
@@ -95,11 +95,11 @@ describe(
 					const mockStats = new Stats({
 						children: [
 							{
-								getStats: () =>
```
Comment (ooflorent) [inline test/Stats.unittest.js]: Please revert this change

File: test/Stats.unittest.js [hunk 2]
```diff
@@ -174,6 +180,16 @@ describe(
 					warnings: []
 				});
 			});
+
+			it('Given showErrors it should return error with its stack', () => {
+				const fakeError = new Error('FAKE STACK');
```
Comment (ooflorent) [inline test/Stats.unittest.js]:
```suggestion
				const fakeError = new Error("FAKE STACK");
```

File: test/Stats.unittest.js [hunk 3]
```diff
@@ -174,6 +180,16 @@ describe(
 					warnings: []
 				});
 			});
+
+			it('Given showErrors it should return error with its stack', () => {
```
Comment (ooflorent) [inline test/Stats.unittest.js]:
```suggestion
			it("returns an error with its stack when errorDetails is true", () => {
```

File: test/Stats.unittest.js [hunk 4]
```diff
@@ -174,6 +180,16 @@ describe(
...
+				const fakeError = new Error('FAKE STACK');
+				fakeCompilation.errors = [fakeError];
+				const mockStats = new Stats(fakeCompilation);
+
```
Comment (ooflorent) [inline test/Stats.unittest.js]:
```suggestion
```

File: test/Stats.unittest.js [hunk 5]
```diff
@@ -134,8 +134,10 @@ describe(
 			});
 		});
 		describe("toJson", () => {
-			it("returns plain object representation", () => {
-				const mockStats = new Stats({
+			let fakeCompilation;
+			beforeEach(() => {
+
```
Comment (ooflorent) [inline test/Stats.unittest.js]:
```suggestion
```

File: lib/Stats.js [hunk 6]
```diff
@@ -316,6 +316,9 @@ class Stats {
 			if (showErrorDetails && e.missing) {
 				text += e.missing.map(item => `\n[${item}]`).join("");
 			}
+			if (showErrorDetails && e.stack) {
```
Comment (sokra) [inline lib/Stats.js]: Some errors already include the stack trace in the `details`. Maybe it makes sense to only print the stack when no details are set.

```suggestion
			if (showErrorDetails && e.stack && !e.details) {
```

Comment (sokra) [reply lib/Stats.js]: The stack also include the `e.message` so avoid printing it twice in this case.

File: test/Stats.unittest.js [hunk 7]
```diff
@@ -174,6 +180,16 @@ describe(
...
 			});
+
+			it('Given showErrors it should return error with its stack', () => {
+				const fakeError = new Error('FAKE STACK');
+				fakeCompilation.errors = [fakeError];
+				const mockStats = new Stats(fakeCompilation);
+
+				const result = mockStats.toJson({ errorDetails: true });
+				
+				expect(result.errors[0]).toEqual(`${fakeError.message}\n${fakeError.stack}`);
```
Comment (sokra) [inline test/Stats.unittest.js]:
```suggestion
				expect(result.errors[0]).toEqual(fakeError.stack);
```

stack already include the message.

File: lib/Stats.js (hunk 6 above)
Comment (shairez) [reply lib/Stats.js]: Makes sense

File: test/Stats.unittest.js (hunk 7 above)
Comment (shairez) [reply test/Stats.unittest.js]: LGTM

File: test/Stats.unittest.js (hunk 5 above)
Comment (shairez) [reply test/Stats.unittest.js]: @ooflorent Good catch. I've pushed an update with all of these fixes

**Comment by sokra** [comment]
This has been added in webpack 5 (next branch)

**Comment by shairez** [comment]
Thanks @sokra !

**Comment by shairez** [comment]
> /cc @shairez friendly ping

@evilebottnawi Thanks for the ping, 
@sokra what's the next step? should I close this one? isn't it merged to 5?

**Comment by vankop** [comment]
fixed in #8425
